simpl/rest.py

```diff
@@ -23,6 +23,77 @@
 STANDARD_QUERY_PARAMS = ('offset', 'limit', 'sort', 'q', 'facets')
 
 
+def One(schema=str):
+    """Expect the input sequence to contain a single value.
+
+    :keyword schema:
+        Custom schema to apply to the input value. Defaults to just string,
+        since this is designed for query params.
+    """
+
+    def f(v):
+        """"""
+        [value] = v
+        return schema(value)
+    return f
+
+
+def Many(schema=str):
+    """"""
+    def f(v):
+        """"""
+        return [schema(x) for x in v]
+    return f
+
+
+def schema(body_schema=None, body_required=False, query_schema=None,
+           content_types=None, default_response=None):
+    """Decorator to parse and validate API body and query string.
+
+    This decorator allows one to define the entire 'schema' for an API
+    endpoint.
+
+    :keyword body_schema:
+
+    :keyword body_required:
+
+    :keyword query_schema:
+
+    :keyword content_types:
+
+    :keyword default_response:
+    """
+    if not content_types:
+        content_types = ['application/json']
+    if not all('json' in t for t in content_types):
+        raise NotImplementedError("Only 'json' body supported.")
+
+    def deco(func):
+        """Return a decorated callable."""
+        def wrapped(*args, **kwargs):
+            """ """
+
+            # validate the request body per the schema (if applicable):
+            body = bottle.request.json
+            if body_required and not body:
+                bottle.abort(400, 'Call body cannot be empty')
+            if body is None:
+                body = default_response
+            if body_schema:
+                try:
+                    body = body_schema(body)
+                except Exception as exc:
+                    bottle.abort(400, str(exc))
+
+            # validate the query string per the schema (if application):
+            query = bottle.request.query
+            validated_query = query_schema(query.dict)
```

Comment:
This will return a 500. We can catch an error and return a 400 with the useful info. For example, in the body call, you get an error like `extra keys not allowed @ data['foo']`, especially if the callable is a voluptuous schema (catch MultipleInvalid). But just in case it is not, we should also look into catching Value, Type, and IndexErrors.